tests/compiler/NonNullable.ts

```diff
@@ -0,0 +1,21 @@
+type nullableString = string | null;
+
+assert(nameof<NonNullable<u32>>() == nameof<u32>());
+assert(nameof<NonNullable<nullableString>>() == nameof<string>());
+assert(nameof<NonNullable<string[] | null>>() == nameof<string[]>());
+
+function assertNonNull<T>(t: T): void {
+  assert(!isNullable<T>(), "T cannot be null");
+  assert(t != null);
+}
+
+function safetyCheck<A>(t: A): void {
+  if (t != null) {
+    assertNonNull(<NonNullable<A>>t);
+  }
+}
+
+let z: nullableString = "z";
+
+assertNonNull<NonNullable<nullableString>>(z!);
+safetyCheck(z);
```

Comment:
It is still not obvious to me - will there be a compilation error in the following case?
```ts
function foo(t: NonNullable<string>): void {
}

foo(null); // compile time error?
```
and 
```ts
let boo: NonNullable<string | null>; // compile time error?
```

Reply:
The first would fail to compile because above it's already shown that `NonNullable<string> == string`. And the second would be the same as `let boo: string`, which does compile see [here](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1ub25lJnJ1bnRpbWU9c3R1YgovKiogQ2FsY3VsYXRlcyB0aGUgbi10aCBGaWJvbmFjY2kgbnVtYmVyLiAqLwpleHBvcnQgZnVuY3Rpb24gZmliKG46IGkzMik6IGkzMiB7CiAgdmFyIGEgPSAwLCBiID0gMQogIGlmIChuID4gMCkgewogICAgd2hpbGUgKC0tbikgewogICAgICBsZXQgdCA9IGEgKyBiCiAgICAgIGEgPSBiCiAgICAgIGIgPSB0CiAgICB9CiAgICByZXR1cm4gYgogIH0KICByZXR1cm4gYQp9CgpsZXQgczogc3RyaW5nOwoKZXhwb3J0IGZ1bmN0aW9uIGZvbygpOiBpMzIgewogIGxldCB4ID0gcy5sZW5ndGgKICByZXR1cm4geAp9CiMhaHRtbAo8dGV4dGFyZWEgaWQ9Im91dHB1dCIgc3R5bGU9ImhlaWdodDogMTAwJTsgd2lkdGg6IDEwMCUiIHJlYWRvbmx5PjwvdGV4dGFyZWE+CjxzY3JpcHQ+CmxvYWRlci5pbnN0YW50aWF0ZShtb2R1bGVfd2FzbSwgeyAvKiBpbXBvcnRzICovIH0pCiAgLnRoZW4oKHsgZXhwb3J0cyB9KSA9PiB7CiAgICBjb25zdCBvdXRwdXQgPSBkb2N1bWVudC5nZXRFbGVtZW50QnlJZCgnb3V0cHV0JykKICAgIGZvciAobGV0IGkgPSAwOyBpIDw9IDEwOyArK2kpIHsKICAgICAgb3V0cHV0LnZhbHVlICs9IGBmaWIoJHtpfSkgPSAke2V4cG9ydHMuZmliKGkpfVxuYAogICAgfQogIH0pCjwvc2NyaXB0Pg==)

Reply:
Btw this:
```ts
const test: NonNullable<string | null> = null;
```
Also will be compile time error, right?